src/constructs/core/stack.ts

```diff
@@ -140,6 +141,25 @@ export class GuStack extends Stack implements StackStageIdentity, GuMigratingSta
 
       this.tryAddRepositoryTag();
     }
+
+    const { node } = this;
+
+    node.addValidation({
```

Comment:
The validation step happens just before synthesis:

![image](https://user-images.githubusercontent.com/836140/168233942-d65d64cc-6579-4b8f-8148-a74a892160ad.png)

See:
  - [App lifecycle](https://docs.aws.amazon.com/cdk/v2/guide/apps.html#lifecycle)
  - The [`constructs` docs](https://github.com/aws/constructs/blob/bd0d7c30c794c9198bd7e866362cd4bc32eeed91/API.md#addvalidationvalidation-)

Reply:
> I think it could also be very useful for ensuring that key best practices are met

I think we could use a combination of this and [`Aspects`](https://aws.amazon.com/blogs/devops/align-with-best-practices-while-creating-infrastructure-using-cdk-aspects/)? I'd like to experiment with placing the [Security Group port 22 check](https://github.com/guardian/cdk/blob/00dfd9d791dba6be96995643ecd439c3441c3095/src/constructs/ec2/security-groups/base.ts#L56-L58) here (in a later PR!) so that all SGs are checked, not just those that extend `GuBaseSecurityGroup`.